File: leetcode/257.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def dfs(self, curnode, path, answer):
        if curnode.left == None and curnode.right == None:
            answer.append(path[:-2])
            return 

        for child in range(2):
            if child == 0 and curnode.left != None:
                self.dfs(curnode.left, path+str(curnode.left.val)+"->", answer)
            elif child == 1 and curnode.right != None:
                self.dfs(curnode.right, path+str(curnode.right.val)+"->", answer)


    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        answer = []
        self.dfs(root, str(root.val)+"->", answer)

        return answer
    

File: leetcode/257.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    void dfs(TreeNode* curnode, string path, vector<string>& answer) {
        if (curnode->left == nullptr and curnode->right == nullptr){
            answer.push_back(path);
            return ;
        }
        for (int i=0; i < 2; i++){
            if (i == 0 && curnode->left!=nullptr){
                dfs(curnode->left, path+"->"+to_string(curnode->left->val), answer);
            }
            else if (i == 1 && curnode->right!=nullptr) {
                dfs(curnode->right, path+"->"+to_string(curnode->right->val), answer);
            }
        }

    }

    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> answer;
        dfs(root, to_string(root->val), answer);
        return answer;
    }
};